[FLINK-21356] Implement incremental checkpointing and recovery using state changelog

rkhachatryan opened the pull request:
Depends on:
- #15200 (writing data)
- #16035 (writing metadata)
-  #15339 (for testing)
-  #16050 (for testing) 
 
Must be rebased before merging. 
Only the last 5 commits are relevant.

## What is the purpose of the change

Add support for checkpointing and recovery using changelog to `ChangelogStateBackend`.
Currently, changelog is written (with #15200) but not used, neither in snapshot nor
during recovery - the underlying backend's snapshot is used.

State materialization will be implemented in FLINK-21357 (for now, the underlying
backend is not snapshotted and the changelog keeps ever growing).

Design doc: https://docs.google.com/document/d/10c6hZsOVxzUjeCLPSDpKGyZOYHi73yd92lCqRs1CyUE/edit?usp=sharing

## Verifying this change

- Existing unit tests which extend `StateBackendTestBase` are used (which test recovery after #16050).
- Integration tests should use `ChangelogStateBackend` after FLINK-21448

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

Myasuka reviewed (commented):
Thanks for creating this PR!
I noticed that the confirm and reset logic for `InMemoryStateChangelogWriter` is empty, are they blocked for the log ownership discussion? And I also need to think more about `notifyCheckpointComplete` and `notifyCheckpointAborted` in your current PR.
Some other comments for changelog state are added in https://github.com/apache/flink/pull/15200

Zakelly commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMapState.java:
```diff
@@ -71,22 +102,66 @@ public boolean contains(UK key) throws Exception {
+                    @Override
+                    public Map.Entry<UK, UV> next() {
+                        return loggingMapEntry(
+                                iterator.next(), changeLogger, changeWriter, currentNamespace);
```
Moreover, I suggest we define a unified writer for ```StateChangeLogger#stateElementRemoved```. For example, define a function here like:
```
private ThrowingConsumer<DataOutputViewStreamWrapper, IOException> getStateChangeWriter(UK key, UV value) {
    return out -> {
        serializeKey(key, out);
        serializeValue(value, out);
    };
}
```
In ```ChangelogMapState#put```, we call it like:
```
    @Override
    public void put(UK key, UV value) throws Exception {
        if (getValueSerializer() instanceof MapSerializer) {
            changeLogger.stateElementChanged(
                    getStateChangeWriter(key, value),
                    getCurrentNamespace());
        } else {
            changeLogger.stateAdded(singletonMap(key, value), getCurrentNamespace());
        }
        delegatedState.put(key, value);
    }
```
As for the inner ```Iterator#next``` of  ```getEntryIterator``` here, we call: 
```suggestion
                                iterator.next(), changeLogger, (entry, out) -> getStateChangeWriter(entry.getKey(), entry.getValue()).accept(out), currentNamespace);
```
It is easier to maintain the output format for a state change.

rkhachatryan replied:
Good idea. I added a commit to #15200 to implement this: e76855e230c997fc1e1507e1a54a6a08b2cf4a31
Is that what you mean?

I'm a bit concerned about the performance, as we'll be creating a lambda on each state change. Which can be significant with Heap State backend.

WDYT?

Zakelly replied:
Yeah, that's what I mean.

Actually, before this change, there is a lambda creation as well (see [here](https://github.com/apache/flink/commit/e76855e230c997fc1e1507e1a54a6a08b2cf4a31#diff-50480dff268c9d790da42d0edb430b91dff4012c861056faf076d465b6080910L86)). I make a simple performance test using JMH (I also attach the source code here [LambdaBench.java.zip](https://github.com/apache/flink/files/6621076/LambdaBench.java.zip)):
```
    @Benchmark
    public void lambdaGetWriter() throws Exception {
        Map.Entry<Integer, Integer> entry = this.entry;
        calling(getWriter(entry.getKey(), entry.getValue()));
    }

    @Benchmark
    public void lambdaPredefinedWriter() throws Exception {
        Map.Entry<Integer, Integer> entry = this.entry;
        calling((out) -> predefinedWriter.accept(entry, out));
    }

    @Benchmark
    public void lambdaInline() throws Exception {
        Map.Entry<Integer, Integer> entry = this.entry;
        calling((out) -> {
            write(entry.getKey(), out);
            write(entry.getValue(), out);
        });
    }

    private BiConsumerWithException<
            Entry<Integer, Integer>, Void, IOException> predefinedWriter =  (entry, out) -> {
        write(entry.getKey(), out);
        write(entry.getValue(), out);
    };
```
It simulate and compare the calling procedure of the lambda before and after the change. The result is here:
```
Benchmark                            Mode  Cnt       Score       Error   Units
LambdaBench.lambdaGetWriter         thrpt   10  336522.176 ±  9458.113  ops/ms
LambdaBench.lambdaInline            thrpt   10  332052.779 ± 10503.985  ops/ms
LambdaBench.lambdaPredefinedWriter  thrpt   10  263978.110 ± 10960.900  ops/ms
```
The ```getWriter``` performs no worse than other ways. However, it's just a simulation and may be different from how it really is. We may do more investigation on the memory consumption, GC and performance after completing the ChangeLogStateBackend.

rkhachatryan replied:
Awesome, thanks a lot!
I agree, we can test performance more thouroughly after completing these parts.

Zakelly commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/TestChangeLoggerKv.java:
```diff
@@ -0,0 +1,74 @@
+import java.io.IOException;
+import java.util.Collection;
+
+class TestChangeLoggerKv<State> implements KvStateChangeLogger<State, String> {
```
Maybe add some correctness test for writing the right key/value when state changed?

rkhachatryan replied:
I've added a couple tests to #15200 (a7a4d1a42fc2658da455d702f141b453361d6cfc). I think correctness should be tested more regorously using more complex tests involving snapshotting and recovery (as it's done in `StateBackendTestBase` descendants).

Zakelly commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLogger.java:
```diff
@@ -0,0 +1,122 @@
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+
```
Better add some javedoc.

rkhachatryan replied:
Sure, 24c8af22f7a4b9dcf93a143188cd7aef53f9e1b2 (also #15200).

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/FunctionDelegationHelper.java:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    static class DelegatingReduceFunction<T>
```
Good point, thanks!

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/StateChangelogHandleStreamImpl.java:
```diff
@@ -29,28 +30,45 @@
 
 /** {@link StateChangelogHandle} implementation based on {@link StreamStateHandle}. */
 @Internal
 public final class StateChangelogHandleStreamImpl
```
`StateChangelogHandleStreamImpl` is using `StreamStateHandle` which can be implemented not only using files (though we currently only have File and ByteStream).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/StateChangelogHandleStreamImpl.java:
```diff
@@ -29,28 +30,45 @@
+        this.size = size;
+    }
+
+    public StateChangelogHandleStreamImpl(
```
Yes, it is used in the actual implementations (e.g. #14839). 
OTH, this handle can not be moved out of `flink-runtime` easily because it needs to be (de)serialized on JM.
That's why I introduced it (and its constructors) independently in this PR (rather than with changelog implementation).
WDYT?

Myasuka replied:
I see, from my point of view, we could add this constructor or another `ChangelogStateHandleStreamImpl#of(List<Tuple3<StreamStateHandle, Long, Long>> sorted, KeyGroupRange keyGroupRange)` method in #14839 when needed.

rkhachatryan replied:
Fair enough, I'll remove it and introduce in the subsequent PR.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/inmemory/InMemoryStateChangelogHandle.java:
```diff
@@ -70,11 +72,26 @@ public KeyGroupRange getKeyGroupRange() {
+    }
+
+    public long getFrom() {
+        return ((SequenceNumber.GenericSequenceNumber) from).number;
```
The idea is to be have different implementations of the changelog, potentially with different sequencing. For example, once could include an epoch and a number inside that epoch.

Adding `long number()` to `SequenceNumber` would expose it's implementation detail.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLogger.java:
```diff
@@ -0,0 +1,122 @@
+                } catch (IOException e) {
+                    ExceptionUtils.rethrow(e);
+                }
+                iterator.remove();
```
I agree and I've changed this in the PR that this one is based on:
https://github.com/apache/flink/pull/15200#discussion_r647618746
(the change should come after merging and rebasing)

Zakelly commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/AggregatingStateChangeApplier.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    protected void applyNameSpaceMerge(DataInputView in) {
+        throw new UnsupportedOperationException();
```
Will implement this later?

rkhachatryan replied:
It actually should be implemented in this PR, thanks for pointing out.

Zakelly commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogReducingState.java:
```diff
@@ -65,16 +80,31 @@ public V get() throws Exception {
     @Override
     public void add(V value) throws Exception {
         delegatedState.add(value);
+        changeLogger.stateUpdated(delegatedState.get(), getCurrentNamespace());
```
For interface semantics, using ```delegatedState.getInternal()``` is better? However, it's no big deal.

rkhachatryan replied:
You are right (it was changed in #15200  and now in master).

Zakelly commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/KvStateStateChangeApplier.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    protected abstract void applyNameSpaceMerge(DataInputView in) throws Exception;
```
```applyNameSpaceMerge``` -> ```applyNamespaceMerge``` or ```applyNamespaceMerged``` ?

Zakelly reviewed (commented):
Thanks for the PR!
Since some interfaces has change in #15200, it's a little difficult to review this PR. I'll check out again after merging and rebasing.

rkhachatryan commented:
Thanks for the review @Zakelly and @Myasuka .
I've addressed your comments and rebased the PR:
(#15200 was merged, #16035 not yet; please ignore the 1st commit `Log metadata`).

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -114,13 +128,33 @@
     private String lastName;
 
+    /** Updated initially on restore and later upon materialization (after FLINK-21356). */
+    private final List<KeyedStateHandle> materialized = new ArrayList<>();
```
this should be changeable right?
remove final?

rkhachatryan replied:
Right (if updated by changing reference it shouldn't be `final` as opposed to `clear` + `addAll`).

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -114,13 +128,33 @@
+    private final List<KeyedStateHandle> materialized = new ArrayList<>();
+
+    /** Updated initially on restore and later cleared upon materialization (after FLINK-21356). */
+    private final List<StateChangelogHandle> nonMaterialized = new ArrayList<>();
```
this is not changeable, just to double confirm?

rkhachatryan replied:
This should only be cleared upon the first materialization.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -245,8 +280,38 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
+                lastUploadedTo);
+        return toRunnableFuture(
+                stateChangelogWriter
+                        .persist(lastUploadedFrom)
```
This returns the Future of StateChangelogHandle that is from 

lastUoloadedFrom till the most recent change

But "nonMaterialized" is not updated accordingly. 

Would this have problems in buildSnapshotResult

rkhachatryan replied:
No, `buildSnapshotResult` receives new changes as a future callback. It then combines them with `materialized` and `nonMaterialized`.
(as commented [above](https://github.com/apache/flink/pull/15420#discussion_r654955828), `nonMaterialized` only contains the state changes from the previous attempt and should be cleared after the first materialization).

rkhachatryan replied:
Yes (all state backend methods are assumed to be called only by the task thread).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java:
```diff
@@ -0,0 +1,154 @@
+ */
+@Internal
+public interface ChangelogStateBackendHandle extends KeyedStateHandle {
+    List<KeyedStateHandle> getMaterializedStateHandles();
+
+    List<StateChangelogHandle> getNonMaterializedStateHandles();
```
Hey Yuan, yes, the order matters here. It should reflect the order of the changes: if the backend had some log state upon recovery, new log state should be ordered after it.

curcur replied:
the order of `List<StateChangelogHandle> getNonMaterializedStateHandles()` matters;

but the order of `List<KeyedStateHandle> getMaterializedStateHandles()` does not, right?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -114,13 +128,33 @@
     private String lastName;
 
+    /** Updated initially on restore and later upon materialization (after FLINK-21356). */
+    private final List<KeyedStateHandle> materialized = new ArrayList<>();
+
+    /** Updated initially on restore and later cleared upon materialization (after FLINK-21356). */
+    private final List<StateChangelogHandle> nonMaterialized = new ArrayList<>();
```
1. Checkpoints are incremental so that one is built on top of another. So both materialized and non-materialized states from the previous checkpoint must be included if they exist. Without restarts, non-materialized part from the previous checkpoint can be collected from the writer. But after recovery, there is a new writer with a new SQNs. So we explicitly include all state from the previous checkpoint.
(Having a new writer for each attempt was a deliberate design decision to avoid split-brain-like issues.)

2. I'm not sure I fully understand the question. The reported state is the conceptually the same, but on TM it's not necessarily built from these two variables. It can be `materialize` + `nonMaterialized` + `newNonMaterialized`

Does this make sense?

I should probably rename `nonMaterialized` to `restoredNonMaterialized`, WDYT?

curcur replied:
Yes, I think that's a good idea!

And also thanks for clarifying, it totally makes sense.

curcur commented:
Mark:

Sync offline: did not see non-materialized part be applied to the delegated state backend. Double confirm

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -245,8 +280,38 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
             throws Exception {
-        return keyedStateBackend.snapshot(
-                checkpointId, timestamp, streamFactory, checkpointOptions);
+        // The range to upload may overlap with the previous one(s). To reuse them, we could store
+        // the previous results either here in the backend or in the writer. However,
+        // materialization may truncate only a part of the previous result and the backend would
+        // have to split it somehow for the former option, so the latter is used.
```
addressed offline.

Zakelly commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/KvStateStateChangeApplier.java:
```diff
@@ -0,0 +1,89 @@
+    @Override
+    public void apply(StateChangeOperation operation, DataInputView in) throws Exception {
+        K key = getState().getKeySerializer().deserialize(in);
+        keyContext.setCurrentKey(key);
```
Will this trigger the ```keySelectionListeners``` in ```AbstractKeyedStateBackend```? Any side effect?

rkhachatryan replied:
Good question.
This can trigger heap incremental TTL cleanup, if enabled. That shouldn't have correctness but may have performance implications. I think this can be addressed separately though.

Zakelly reviewed (commented):
@rkhachatryan Thanks for the update. I made another pass, PTAL.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java:
```diff
@@ -0,0 +1,154 @@
+            return nonMaterialized;
+        }
+
+        public ChangelogStateBackendHandleImpl withDelta(ChangelogStateHandle delta) {
```
Yes, it's a leftover after a refactoring.
I'll remove it, sorry.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/inmemory/InMemoryChangelogStateHandle.java:
```diff
@@ -17,49 +17,50 @@
-        return CloseableIterator.fromList(
-                changes.entrySet().stream().flatMap(this::mapEntryToChangeStream).collect(toList()),
-                change -> {});
+        return changes.stream().mapToInt(change -> change.getChange().length).sum();
```
Good point.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -116,16 +140,41 @@
+    private final List<ChangelogStateHandle> restoredNonMaterialized = new ArrayList<>();
+
+    @Nullable private SequenceNumber lastUploadedFrom;
+    @Nullable private SequenceNumber lastUploadedTo;
```
Right. I'll add javadoc.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/inmemory/InMemoryStateChangelogWriter.java:
```diff
@@ -61,20 +64,26 @@ public SequenceNumber lastAppendedSequenceNumber() {
     }
```
> the sqn of appended value starts from 1L

This is controlled by the writer, and in `InMemoryStateChangelogWriter` it's `0L`.
The backend then will request to persist changes from `0L` according to `lastUploadedFrom`  = `materializedTo`.
Does this resolve your concern?

Zakelly replied:
It makes sense.
However, it makes me a little confused that sequence number for records in ```InMemoryStateChangelogWriter ``` actually starts from ```1L```, not the ```SequenceNumber.FIRST```. The word FIRST here means an initialized value instead of a valid beginning, right? If so, it does not bother me.

rkhachatryan replied:
That's right and I agree it is confusing.
I'm going to replace `SequenceNumber.FIRST` with a new method `writer.initialSequenceNumber()`. So that writer could use whatever initial SequenceNumber (`0L` in `InMemoryStateChangelogWriter`).

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -358,7 +461,23 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
+
+    private void completeRestore(Collection<ChangelogStateBackendHandle> stateHandles) {
+        if (!stateHandles.isEmpty()) {
+            synchronized (materialized) { // ensure visibility
```
The variables updated in this method (by the task thread) are also read in `buildSnapshotResult`, potentially from a different thread.
Without `synchronized` (or other means to inject memory barrier) I'm afraid those other threads may see not updated values.

Zakelly replied:
IIUC, the ```buildSnapshotResult ``` runs in synchronized stage of a checkpoint, which is already protected by the checkpoint lock.
However, since the ```materialized``` may change later upon materialization, I do agree we add some protection here. According to the JLS (https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5):

> An unlock on a monitor happens-before every subsequent lock on that monitor.

I think we should do ```synchronized (materialized)``` in ```buildSnapshotResult ``` as well to ensure the _happens-before_ relationship.

curcur commented:
In `ChangelogListState.mergeNamespaces` 
you mentioned that
"To prevent duplicated effects on recovery record the operation", Can you elaborate what are the duplicated efforts and why in this particular case order matters?

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -116,16 +140,54 @@
+     * The {@link SequenceNumber} up to which the state is materialized, exclusive. The log should
+     * be truncated accordingly.
+     *
+     * <p>WARN: currently not updated - to be changed in FLINK-21356.
```
This PR targets for FLINK-21356 and field `materializedTo` is still not updated, this really confuse me.

rkhachatryan replied:
Without materialization, the backend is still incomplete and my intention was to make it clear what else and how will change. It's not ideal, but leaving `lastUploadedFrom` never updated (or always updated to 0) would have the same or worse effect IMO.
I can remove it though if you'd like.

Zakelly replied:
@rkhachatryan I thought it's the statement bring some confusion? Maybe 'to be change AFTER FLINK-21356'(or 'IN' any other FLINK issue) could be better? @Myasuka, is that your concern?

Myasuka replied:
I think `materializedTo` would be updated in later PR (not FLINK-21356). If so, `to be changed in FLINK-21356` seems not correct.

rkhachatryan replied:
Oh, sure, it should be FLINK-21357 not FLINK-21356 🙃
I'll update the comment if you're fine with that.
Sorry for the confusion.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -116,16 +140,54 @@
+    /** Updated initially on restore and later upon materialization (after FLINK-21356). */
+    private final List<KeyedStateHandle> materialized = new ArrayList<>();
+
+    /** Updated initially on restore and later cleared upon materialization (after FLINK-21356). */
```
What does `after FLINK-21356` means? If I understand correctly, this `restoredNonMaterialized` would be cleared upon materialization which might be implemented in FLINK-21357, right? If so, why not change the doc to `in FLINK-21357`?
The same comment for `materialized` field.

rkhachatryan replied:
Right, I'll change the comment accordingly.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendLogApplier.java:
```diff
@@ -0,0 +1,176 @@
+import static org.apache.flink.state.changelog.restore.FunctionDelegationHelper.delegateReduceFunction;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Applies {@link StateChange}'s to a {@link ChangelogKeyedStateBackend}. */
```
A typo, I think the correct javadoc should be `Applies {@link StateChange}s to a {@link ChangelogKeyedStateBackend}.`

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendLogApplier.java:
```diff
@@ -0,0 +1,176 @@
+                changelogBackend,
+                in,
+                classLoader,
+                new ChangelogApplierFactoryImpl());
```
I think a singleton instance of `ChangelogApplierFactoryImpl` should be better here.

rkhachatryan replied:
I agree, will change it, thanks.

rkhachatryan commented:
Thanks for your feedback, I've addressed your comments, PTAL.

@Myasuka
> I noticed that the confirm and reset logic for InMemoryStateChangelogWriter is empty, are they blocked for the log ownership discussion?

No, `InMemoryStateChangelogWriter` just doesn't have to do anything as it keeps everything in memory. So persist() returns handles for the appended changes regardless of any JM decisions.

@curcur 
> you mentioned that "To prevent duplicated effects on recovery record the operation", Can you elaborate what are the duplicated efforts and why in this particular case order matters?

It was 02e33f112973f2d1a228ebdd6a7f48fe9b2aa0f4 (`[hotfix][state/changelog] Merge namespaces after logging it`) and I see it was a mistake:
I assumed that `changeLogger.namespacesMerged` uses the underlying state, which is not the case. I'll drop this commit.

Myasuka commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -369,7 +509,73 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
+        ChangelogState state;
+        switch (type) {
+            case KEY_VALUE:
+                state = (ChangelogState) keyValueStatesByName.get(name);
```
I think there exists something wrong here. `keyValueStatesByName` which is like `AbstractKeyedStateBackend#keyValueStatesByName`, is only updated during `KeyedStateBackend#getOrCreateKeyedState`. 

However, this place is not somewhere we restore and build the keyed state backend but only for the first user access. For RocksDB state backend, `kvStateInformation` is the actual place to store restored state. And for heap keyed state backend, `registeredKVStates` is the place to store restored data.

That is to say, we cannot get existing state when try to apply state changes during restore. Moreover, we also need to consider state migration, the state changes should be appiled by previous state descriptor (mainly the value seralizer) not current state descriptior (as they might not be the same).

rkhachatryan replied:
`ChangelogKeyedStateBackend` differs a bit from other backends here. On recovery,
1. delegated backends are created as usual (with metadata `M1`)
1. `ChangelogBackendLogApplier` reads metadata `M2` in `restoreKvMetaData()` and calls `changelogBackend.getOrCreateKeyedState()` (`M2` replaces `M1` in delegated backend)
1. After that, `ChangelogBackendLogApplier.applyDataChange()` calls `changelogBackend.getExistingState()` to apply data changes
1. User program may call `changelogBackend.getOrCreateKeyedState()` with metadata `M3`

So `changelogBackend.keyValueStatesByName` **is** updated on recovery (in step 2).

---

However, state migration won't currently work as you pointed out.

I see the only way to implement it by using delegating metadata and serializers, similar to delegating functions in this PR:
- on step 2, wrap obtained serializers in `changelogBackend.createInternalState()` and pass this to the `delegatedBackend`
- on step 4, upgrade metadata and replace the wrapped serializers with the upgraded ones

So the metadata of the delegated backend won't be upgraded explicitly, but effictively it will start using the new serializers.

WDYT? (it's probably easier to discuss it offline)

Other approaches don't seem to work:
- If we simply call `delegatedBackend.createInternalState` to upgrade metadata, then already created states won't receive the updated serializers
- If we replace existing internal states with new ones, we can end up with resource leaks (and potentially non-flushed changes)
- (we do need to create internal delegated states during recovery, so that we can apply changes by calling `put`/`add`/`clear` on them)

Myasuka replied:
I think we should not call `changelogBackend.getOrCreateKeyedState()` in step2 to put states within delegated backend. And I think delegating state descritptor which includes serializer and function could help here. We would not let delegated backend to call `getOrCreateKeyedState` during restore.

rkhachatryan replied:
> I think we should not call changelogBackend.getOrCreateKeyedState() in step2 to put states within delegated backend.

But how will we apply state changes then? When applying changelog, delegated backend **has** to have internal states created.
Do you mean we should call some other method but still have internal states created in step 2?

> We would not let delegated backend to call getOrCreateKeyedState during restore.

`delegatedBackend.getOrCreateKeyedState` is actually never called. Only `delegatedBackend.createInternalState` is called.

> And I think delegating state descritptor which includes serializer and function could help here.

In the end, internal states have `final` references to serializers. So at least serializers should be delegating.
I think state migration deserves a separate ticket and a PR, which can be applied after this one, WDYT?

Myasuka replied:
Let me take `RocksDBKeyedStateBackend` for example.

Operations | Restoring state | User starts to call getOrCreateState for the 1st time
--|------------ | -------------
Normal RocksDBKeyedStateBackend|Restore from checkpoingts, not visiable for users (not stored in `keyValueStatesByName`) | Might trigger state migration and then visiable for users (stored in `keyValueStatesByName`)
Changelog (RocksDB) statebackend|Restore from checkpoingts, visiable for users and then apply state changes ( (stored in `keyValueStatesByName`) | Just get from `keyValueStatesByName` and cannot trigger state migration.

I think the first difference is that shall we make the state visiable for users during restoring for changelog state backend? From my point of view, we could still keep them not visiable so that we can then trigger state migration for the 1st user access.
I agree that we could make state migration as a separate ticket, but I think we should first discuss whether we should make previous state visiable in `keyValueStatesByName` and that could be an easy fix in this PR.

If talking about state migration, I think we could have two choices:
1. Apply state changes during restore. And then migrate state in the 1st user access. This might need data in changelog write twice.
2. Keep the state changes until user access the state during the 1st time. Then we could apply state changes with possible state migration. This could avoid uncessary re-write, but we need to keep those persist state changes.

Maybe we could have basic discussion here for these choices and leave TODO in this PR if necessary.

rkhachatryan replied:
To my understanding, the purpose of `keyValueStatesByName` is to have 
1. exactly-once state table creation (in `createInternalState`)
2. exactly-once metadata update and state migration (in `tryRegisterKvStateInformation`) 

To implement (2), we'll probably introduce a special `changelogBackend.getOrCreateStateForRecovery()`. 
And then both methods will provide (1) - with the help `keyValueStatesByName` map;
and only in the current `getOrCreateState()` will provide (2) - by using some `delegatingSerializers` map for example.
So `keyValueStatesByName` should be updated anyways.

Regarding the approaches, lazy recovery (option 2) seems much more complicated. And because most snapshots I think will have the materialized part (which will be bigger), it's re-write will dominate changelog re-write. So I think it doesn't worth it.

WDYT?

rkhachatryan replied:
I've created FLINK-23143 and will add a corresponding `TODO` in `ChangelogKeyedStateBackend.getOrCreateKeyedState` as agreed offline.

Thanks a lot for bringing this up and for the discussion!

Myasuka reviewed (commented):
Thanks for the update!

As some dependent commits have been merged, I take another look over the whole PR. Please take a look at my comments.

curcur commented:
Another suggestion: which might be a bit picky but will help greatly readability and error prune in the future:

For each 
`XXXStateChangeApplier` and `ChangelogXXXState` (like `ListStateChangeApplier` and `ChangelogListState`)

Is it possible to make a more consistent naming? like this in `ListStateChangeApplier`

 @Override
    protected void applyStateElementAdded(DataInputView in) throws Exception {
        TypeSerializer<List<T>> valueSerializer = state.getValueSerializer();
        if (valueSerializer instanceof ListSerializer) {
            state.add(((ListSerializer<T>) valueSerializer).getElementSerializer().deserialize(in));
        } else {
            state.addAll(valueSerializer.deserialize(in));
        }
    }

correspond to `ChangelogListState`
    @Override
    public void add(V value) throws Exception {
        delegatedState.add(value);
        changeLogger.valueElementAdded(
                w ->
                        ((ListSerializer<V>) getValueSerializer())
                                .getElementSerializer()
                                .serialize(value, w),
                getCurrentNamespace());
    }

===
I am asking because there are a lot of similar methods like update, update internal, e.t.c

Right now, it is difficult to map them directly and hard to check which should be supported/unsupported; missed or redundant.

rkhachatryan commented:
Thanks for the feedback and suggestion @curcur, I agree that this part is hard to follow currently.

I simplified appliers by pushing down `apply(operation)`, PTAL
(having consistent names for all methods seems impossible because different implementations use different state methods)

curcur commented:
Thanks @rkhachatryan for the quick amends!

Another question (sorry does not find where the change was made) is about `InMemoryChangelogStateHandle`. 
Why `getKeyGroupRange()` is not supported in `InMemoryChangelogStateHandle`? For testing purposes, KeyGroupRange part is also important right?

curcur commented:
I think I've made similar comments and asked similar questions before, now I've forgotten it again, LOL; so it either worth making it POJO or at leave a comment.

**Need to discuss offline**

1. In `ChangelogStateHandleStreamImpl`, what is 
`List<Tuple2<StreamStateHandle, Long>> handlesAndOffsets`, what is the offset? and there is an additional added field `size`

2. who will call "registerSharedStates", How is `SharedStateRegistry` related to the JM owned shared counter in the current model?

curcur commented:
Need to discuss offline:

I want to understand a bit more about the relationship between

`ChangelogStateBackendHandle`

`ChangelogStateHandle` and normal `KeyedStateHandle`

1). Each ChangelogStateHandle one task?
2). ChangelogStateBackendHandle multiple tasks?
3). shared registry should be registered through `ChangelogStateBackendHandle`? then why "ChangelogStateHandle" also implement the register function. In what case it is used?

curcur commented:
In `ChangelogKeyedStateBackend.completeRestore`, where access to `materialized` is synchronized (Again, I can not find where the code is added)..

While I am fine to synchronize on this field, what I want to point out is at the point where the complete restore is called, the keyedstatebackend is not finish building; which means you would neither update the materialized data nor building snapshot.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/AbstractKeyedStateBackend.java:
```diff
@@ -53,7 +53,8 @@
                 CheckpointListener,
-                TestableKeyedStateBackend<K> {
+                TestableKeyedStateBackend<K>,
+                InternalKeyContext<K> {
```
Sure, it is used on recovery to change the current key, keyGroup before applying state changes.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/KvStateStateChangeApplier.java:
```diff
@@ -0,0 +1,89 @@
+import org.apache.flink.runtime.state.internal.InternalKvState;
+import org.apache.flink.state.changelog.StateChangeOperation;
+
+abstract class KvStateStateChangeApplier<K, N> implements StateChangeApplier {
```
Yes, it's a typo, sorry about that :)

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java:
```diff
@@ -151,10 +160,11 @@ private void logMetaIfNeeded() throws IOException {
+                    wrapper.writeByte(op.getCode());
                     // todo: optimize in FLINK-22944 by either writing short code or grouping and
                     // writing once (same for key, ns)
                     wrapper.writeUTF(metaInfo.getName());
+                    wrapper.writeByte(stateType.getCode());
```
Yes, different state types are written to the same log. So they need to be differentiated on recovery (each change).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/inmemory/InMemoryStateChangelogStorage.java:
```diff
@@ -19,14 +19,14 @@
-        implements StateChangelogWriterFactory<InMemoryStateChangelogHandle>, Serializable {
+/** An in-memory (non-production) implementation of {@link StateChangelogStorage}. */
+public class InMemoryStateChangelogStorage
+        implements StateChangelogStorage<InMemoryStateChangelogHandle>, Serializable {
```
Neither of them actually should be Serializable.
It was needed because instantiation was made in `StateBackendLoader`, but that was revisited (see FLINK-21804).
I'm going to remove `implements Serializable` from both.

curcur commented:
Sync up offline:

`FunctionDelegationHelper`

1. can we use it to prevent simulating add by update.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ValueStateChangeApplier.java:
```diff
@@ -0,0 +1,73 @@
+    @Override
+    protected void applyStateUpdateInternal(DataInputView in) throws Exception {
+        applyStateUpdate(in);
+    }
```
I removed this method while addressing your above comment (value state doesn't have updateInternal so it shouldn't be here).

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ValueStateChangeApplier.java:
```diff
@@ -0,0 +1,73 @@
+    @Override
+    protected void applyStateElementAddedOrUpdated(DataInputView in) throws Exception {
+        state.update(state.getValueSerializer().deserialize(in));
+    }
```
Right, removed while addressing one of your above comments.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ListStateChangeApplier.java:
```diff
@@ -0,0 +1,90 @@
+        } else {
+            state.addAll(valueSerializer.deserialize(in));
+        }
+    }
```
It shouldn't be. If check removed while addressing one of your above comments in 40cae883fa9daa4a33354d391cbaff9f580232b8.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ListStateChangeApplier.java:
```diff
@@ -0,0 +1,90 @@
+            sources.add(state.getNamespaceSerializer().deserialize(in));
+        }
+        state.mergeNamespaces(target, sources);
+    }
```
Agreed, already done that in 40cae883fa9daa4a33354d391cbaff9f580232b8.

rkhachatryan commented:
I've updated the PR, PTAL @Myasuka and @curcur (@Zakelly do you have any other concerns?)

@curcur 
Regarding `completeRestore()`,
I'm not sure I fully understand your point. Is it about calling  before completing recovery of the underlying backend?
But `ChangelogKeyedStateBackend.completeRestore()` is called from `ChangelogBackendRestoreOperation.restore()` - after creating the underlying state backend (in fact, it is a constructor argument).

Regarding `InMemoryChangelogStateHandle.getKeyGroupRange()` - I've added an implementation, thanks for pointing out.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java:
```diff
@@ -0,0 +1,154 @@
+            return nonMaterialized;
+        }
+
+        public ChangelogStateBackendHandleImpl withDelta(StateChangelogHandle delta) {
```
Already [addressed](https://github.com/apache/flink/pull/15420#discussion_r655963823).

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -240,8 +275,38 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
             throws Exception {
-        return keyedStateBackend.snapshot(
-                checkpointId, timestamp, streamFactory, checkpointOptions);
+        // The range to upload may overlap with the previous one(s). To reuse them, we could store
+        // the previous results either here in the backend or in the writer. However,
+        // materialization may truncate only a part of the previous result and the backend would
+        // have to split it somehow for the former option, so the latter is used.
```
Some component has to maintain a data structure allowing range queries for changes (SqN >= ...).
Currently (in #14839/#15322), this is done in Writer with a sorting tree.
But essentially this is an index over the log and can be implemented differently by different stores; and might need some maintanence which is not the backend responsibility.

Zakelly commented:
> I've updated the PR, PTAL @Myasuka and @curcur (@Zakelly do you have any other concerns?)
I have no more concerns :)

Myasuka commented:
I think this PR looks good in general and maybe @curcur could have more discussion on the part of thread model to see whether `synchronized` could be optimized.

curcur commented:
After syncing up offline, I am fine with the current threading model.

rkhachatryan commented:
Thanks a lot for reviewing!

Rebased, will merge as soon as tests pass.
